Objects/frameobject.c

```diff
@@ -159,6 +183,16 @@ frame_setlineno(PyFrameObject *f, PyObject* p_new_lineno)
 
     /* We're now ready to look at the bytecode. */
     PyBytes_AsStringAndSize(f->f_code->co_code, (char **)&code, &code_len);
+
+    /* The trace function is called with a 'return' trace event after the
+     * execution of a yield statement. */
+    assert(f->f_lasti != -1);
+    if (code[f->f_lasti] == YIELD_VALUE || code[f->f_lasti] == YIELD_FROM) {
```

Comment:
Is there a need to check for RETURN_VALUE as in your original patch?

Reply:
The only function called by ``_PyEval_EvalFrameDefault()`` that may allow a jump to another line is ``maybe_call_line_trace()`` because ``stack_pointer`` is saved on ``f->f_stacktop`` before the call and there is a ``JUMPTO(f->f_lasti)`` after the call. All the other trace functions called within ``_PyEval_EvalFrameDefault()`` are called with ``f->f_stacktop == NULL`` except after the handling of the ``YIELD_FROM`` and ``YIELD_VALUE`` opcodes. So by forbidding jumps in ``frame_setlineno()`` when ``f->f_stacktop == NULL`` we handle all the cases, including the ``RETURN_VALUE`` opcode, except the yield statements. The check for ``RETURN_VALUE`` in my original patch was overkill.